add support for Lights (ZwaveDimmer)

marcelveldt opened the pull request:
Basic light support (non-color dimmer only). Color support to be added later.

Also a includes a small fix for displaying the power_consumption as  state attribute if a device supports this.

marcelveldt committed:
add support for Lights (ZwaveDimmer)

Basic light support (non-color dimmer only). Color support to be added later.

Also a includes a small fix for displaying the power_consumption as  state attribute if a device supports this.

balloob commented on custom_components/zwave_mqtt/light.py:
```diff
@@ -0,0 +1,123 @@
+    async_dispatcher_connect(hass, "zwave_new_light", async_add_light)
+
+
+# pylint: enable=unnecessary-pass,unused-argument
```
Remove this.

balloob commented on custom_components/zwave_mqtt/light.py:
```diff
@@ -0,0 +1,123 @@
+        async_add_entities([light])
+
+    async_dispatcher_connect(hass, "zwave_new_light", async_add_light)
+
```
```suggestion

    await hass.data[DOMAIN][config_entry.entry_id]["mark_platform_loaded"]("switch")
```

marcelveldt replied:
Can you explain the difference ?

marcelveldt replied:
got it

balloob commented on custom_components/zwave_mqtt/entity.py:
```diff
@@ -157,7 +158,11 @@ def device_info(self):
+        attrs = {
+            const.ATTR_NODE_ID: self.values.primary.node.node_id,
+        }
+        if self.power_consumption is not None:
```
We are calculating `power_consumption` twice, since it's a calculated property. 

Also, you're adding this to the generic Z-Wave entity. Is this on purpose ?

marcelveldt replied:
In what other place is the power_consumption calculated ? 
And yes, Added it on purpose to the generic entity, so it will be picked up by every (future) device implementation. Currently supported by switches and lights as far as I can see.

Looks like the previous/current z-wave implementation uses the same approach.

marcelveldt replied:
@balloob any word on this ?

balloob commented:
Please include a test. See #39 for an example.

balloob commented:
Don't forget to update the README with that lights are supported too

marcelveldt commented:
All set!

marcelveldt commented:
resolves #1

cgarwood commented:
@balloob the old component would put a power_consumption attribute on every entity if the device offered that information. Is it still preferred for that to be an attribute of a `light`, `switch`, etc, or should they just be broken out into their own sensors?

As this PR stands it matches the power_consumption handling of the existing component, which I'm fine with, but wasn't sure what would be preferred from the HA core side of things.

balloob commented:
Well we can't graph attributes in Home Assistant. Better break it out and have it be combined at device level

marcelveldt commented:
Alright, will do. In fact, there's already a sensor for the power consumption in place: electricity_w
So, all agree to just remove the entire attribute from the entity ?

Also, there is node_id as entity attribute. Keep that one or can we make it a normal attribute instead of state attribute (so it won't be shown in frontend cards) ?

marcelveldt committed:
remove power_consumption state attribute

power_consumption was historically added as state attribute to devices that supported it. Now it has it's own sensor (device_electricity_w)

marcelveldt commented:
@balloob @cgarwood all set now. Removed the power_consumption attr but I left the node_is attr in.

balloob commented on custom_components/zwave_mqtt/entity.py:
```diff
@@ -15,6 +15,8 @@
 
 _LOGGER = logging.getLogger(__name__)
 
+ATTR_POWER = "power_consumption"
```
We can remove this.

balloob commented on custom_components/zwave_mqtt/light.py:
```diff
@@ -0,0 +1,125 @@
+
+    await hass.data[DOMAIN][config_entry.entry_id]["mark_platform_loaded"]("light")
+
+    return True
```
Drop this. It's not necessary.

marcelveldt replied:
It's in line with the switch component. What does it do ?

balloob replied:
It's been long gone 😉 #40

It doesn't do anything.

balloob commented on custom_components/zwave_mqtt/light.py:
```diff
@@ -0,0 +1,125 @@
+        self._supported_features = None
+        self.value_added()
+
+    def value_added(self):
```
Please annotate with `@callback` so it's cleaer it's async friendly.

balloob reviewed (approved):
Ok to merge when final comments addressed

marcelveldt committed:
follow up fixes

prefer target over level if exists on light
fix a few typos and styling stuff

balloob commented on custom_components/zwave_mqtt/light.py:
```diff
@@ -61,6 +60,8 @@ def value_added(self):
     @property
     def brightness(self):
         """Return the brightness of this light between 0..255."""
+        if self.values.get("Target") is not None:
```
Can we test `self.values.target is not None` ?

marcelveldt replied:
not sure, isn't that a nullptr if it doesn't exist ? It's an optional attribute, not present on many lights.

marcelveldt committed:
add "in" operator

to lookup value exists in entity values

marcelveldt committed:
lookup value in dict with get
